node/pkg/reporter/node.go

```diff
@@ -0,0 +1,199 @@
+package reporter
+
+import (
+	"context"
+	"crypto/rand"
+	"errors"
+	"math/big"
+	"os"
+	"time"
+
+	"bisonai.com/orakl/node/pkg/db"
+	"bisonai.com/orakl/node/pkg/raft"
+	"bisonai.com/orakl/node/pkg/utils"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	TOPIC_STRING    = "orakl-offchain-aggregation-reporter"
+	LEADER_TIMEOUT  = 5 * time.Second
+	MAX_RETRY       = 3
+	MAX_RETRY_DELAY = 500 * time.Millisecond
+	FUNCTION_STRING = "batchSubmit(string[] memory _pairs, int256[] memory _prices)"
+
+	GET_LATEST_GLOBAL_AGGREGATES_QUERY = `
+		SELECT ga.name, ga.value, ga.round, ga.timestamp
+		FROM global_aggregates ga
+		JOIN (
+			SELECT name, MAX(round) as max_round
+			FROM global_aggregates
+			GROUP BY name
+		) subq ON ga.name = subq.name AND ga.round = subq.max_round;`
+)
+
+type Reporter struct {
+	Raft     *raft.Raft
+	TxHelper *utils.TxHelper
+
+	lastSubmissions map[string]int64
+	contractAddress string
+}
+
+type GlobalAggregate struct {
+	Name      string    `db:"name"`
+	Value     int64     `db:"value"`
+	Round     int64     `db:"round"`
+	Timestamp time.Time `db:"timestamp"`
+}
+
+func New(ctx context.Context, h host.Host, ps *pubsub.PubSub) (*Reporter, error) {
+	encryptedTopic, err := utils.EncryptText(TOPIC_STRING)
+	if err != nil {
+		return nil, err
+	}
+
+	topic, err := ps.Join(encryptedTopic)
+	if err != nil {
+		return nil, err
+	}
+
+	raft := raft.NewRaftNode(h, ps, topic, 100, LEADER_TIMEOUT)
+	txHelper, err := utils.NewTxHelper(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	contractAddress := os.Getenv("SUBMISSION_PROXY_CONTRACT")
+	if contractAddress == "" {
+		return nil, errors.New("SUBMISSION_PROXY_CONTRACT not set")
+	}
+
+	reporter := &Reporter{
+		Raft:            raft,
+		TxHelper:        txHelper,
+		contractAddress: contractAddress,
+		lastSubmissions: make(map[string]int64),
+	}
+	reporter.Raft.LeaderJob = reporter.leaderJob
+	reporter.Raft.HandleCustomMessage = reporter.handleCustomMessage
+
+	return reporter, nil
+}
+
+func (r *Reporter) Run(ctx context.Context) {
+	r.Raft.Run(ctx)
+}
+
+func (r *Reporter) leaderJob() error {
+	ctx := context.Background()
+	failureTimout := 50 * time.Millisecond
+	for i := 0; i < MAX_RETRY; i++ {
+		n, err := rand.Int(rand.Reader, big.NewInt(100))
```

Comment:
@nick-bisonai Minor, but it would be nice if we have a separate function that generates delay for us, without mixing it too much with the rest of the reporter process. It would also then help with switching between different timeout logic.